zc_common/remote_resource/request.py

```diff
@@ -80,19 +80,30 @@ def add_items_from_data(self):
         map(lambda x: self.append(RemoteResourceWrapper(x)), self.data)
 
 
-def get_route_from_fk(resource_type, pk=None):
-    """Gets a fully qualified URL for a given resource_type, pk"""
-    routes = requests.get(zc_settings.GATEWAY_ROOT_PATH).json()
-
-    for route in routes.iterkeys():
-        if 'resource_type' in routes[route] and routes[route]['resource_type'] == resource_type:
-            if isinstance(pk, (list, set)):
-                expanded = '{}?filter[id__in]={}'.format(expand(route, {}), ','.join([str(x) for x in pk]))
-            else:
-                expanded = expand(route, {'id': pk})
-            return '{0}{1}'.format(routes[route]['domain'], expanded)
-
-    raise RouteNotFoundException('No route for resource_type: "{0}"'.format(resource_type))
+class RouteManager(object):
```

Comment:
@birkholz What is the practical reason of using a RouteManager?

1. Initially, changing the gateway routes required almost no changes to existing services. But now, each time gateway routes are updated services must be restarted if they want to use the changes introduced.
2. We have a plan to change the way we register routes to allow services to rather register their routes with gateway automatically at boot time (instead of hard coding them in routes.js). We would not be able to take advante of that gateway change with this RouteManager (obviously because of the caching).
3. When running unit tests, I am restricted to populate the cache with initial routes. I don't see how that is any helpful.
4. A much less-restrictive approach would be to cache the first initial request to the gateway (and not initiating the request before the service even attempt to make one). At least this way unit tests will pass without additional changes. Even then, refreshing the cache maybe after a few minutes would be a better way than hitting the gateway once and never do so again.